content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,258 @@
+In this example it means that the Jenkins node contains the attacked FPGA of such type.*/
+node ("linux && $fpgaId") {  
+  stage "Blast bitfile"
+  git url:"http://github.com/oleg-nenashev/pipeline_hw_samples";
+  def artifact='target/image_'+fpgaId+'.bit';
+  echo "Using ${artifact}" 
+  unstash 'bitfile'
+  sh "make FPGA_TYPE=$fpgaId impact"
+  
+  /* We run automatic tests.
+  Then we report test results from the generated JUnit report. */
+  stage "Auto Tests"
+  sh "make FPGA_TYPE=$fpgaId tests"
+  sh "perl scripts/convertToJunit.pl --from=target/test-results/* --to=target/report_${fpgaId}.xml --classPrefix=\"myprj-${fpgaId}.\"";
+  step([$class:"JUnitResultArchiver", testResults:"target/report_${fpgaId}.xml"])
+  
+  stage "Finalization"
+  sh "make FPGA_TYPE=$fpgaId flush_fpga"
+  hipchatSend("${imageId} testing has been completed");
+}
+```
+
+As you may see, the pipeline script mostly consists of various calls of command-line tools via the _sh()_ command. 
+All EDA tools provide great CLIs, so we do not need special plugins in order to invoke common operations from Jenkins.
+
+WARNING: Disclaimer. 
+Makefile above is a sample stuff for demo purposes.
+Never write such makefiles.
+
+It is possible to continue expanding the pipeline in such way.
+link:https://github.com/jenkinsci/pipeline-examples[Pipeline Examples]
+contain examples for common cases: build parallelization, code sharing between pipelines, error handling, etc.
+
+## Lessons learned
+
+During the last 2 years I've tried using Pipeline for Hardware test automation several times.
+The first attempts were not very successful, but the ecosystem has been evolving rapidly. 
+There are still several missing integrations, but I feel Pipeline becomes a really powerful tool.
+
+I would like to mention the following improvement areas:
+
+* *Shared resource management across pipelines*. It can be done by the incoming Pipeline integration in the 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Lockable+Resources+Plugin[Lockable Resources plugin] 
+(link:https://issues.jenkins-ci.org/browse/JENKINS-30269[JENKINS-30269]).
+Another case is integration with 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Throttle+Concurrent+Builds+Plugin[Throttle Concurrent Builds plugin], which is an effective engine for limiting the license utilization in automation infrastructures 
+(link:https://issues.jenkins-ci.org/browse/JENKINS-31801[JENKINS-31801]).
+* *Better support of CLI tools*. 
+EDA tools frequently need a complex environment, which should be deployed on nodes somehow. 
+Integration with 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Custom+Tools+Plugin[Custom Tools Plugin] seems to be the best option, especially in the case of multiple tool versions 
+(link:https://issues.jenkins-ci.org/browse/JENKINS-30680[JENKINS-30680]).
+* *Pipeline package manager* with dependency management, which would allow developing Pipeline libraries and sharing them between teams. 
```

Comment:
This feature is accounted for with the global library and this statement doesn't make it clear how this is relevant to the subject of the article